File: mobile/ios/App/App/PalMirrorPlugin.swift
import Foundation
import Capacitor
import AVFoundation

/// Native front-camera presence check for the "mirror" feature. The mirror's
/// camera feed itself is plain web getUserMedia in the WKWebView (granted via the
/// WKUIDelegate in MainViewController); this plugin only answers the gating
/// question "does this device have a front camera?" definitively and WITHOUT
/// triggering a permission prompt or opening the camera, so the UI can hide the
/// mirror button on devices that have none.
///
/// `hasFrontCamera()` resolves `{ present: Bool }`. Never rejects.
@objc(PalMirrorPlugin)
public class PalMirrorPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "PalMirrorPlugin"
    public let jsName = "PalMirror"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "hasFrontCamera", returnType: CAPPluginReturnPromise),
    ]

    @objc func hasFrontCamera(_ call: CAPPluginCall) {
        // Enumerate every front-position video device type (TrueDepth-only front
        // cameras don't match .builtInWideAngleCamera alone), and fall back to the
        // default front-camera lookup. Device discovery does not require camera
        // authorization, so this answers "does the hardware exist?" before any
        // permission prompt.
        var types: [AVCaptureDevice.DeviceType] = [.builtInWideAngleCamera, .builtInTrueDepthCamera]
        if #available(iOS 15.4, *) { types.append(.builtInLiDARDepthCamera) }
        let session = AVCaptureDevice.DiscoverySession(
            deviceTypes: types,
            mediaType: .video,
            position: .front
        )
        let present = !session.devices.isEmpty
            || AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .front) != nil
        call.resolve(["present": present])
    }
}

File: mobile/android/app/src/main/java/sh/martinez/pal/companion/PalMirrorPlugin.java
package sh.martinez.pal.companion;

import android.content.Context;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraManager;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * Native front-camera presence check for the "mirror" feature. The mirror's
 * camera feed itself is plain web getUserMedia in the WebView (Capacitor's
 * BridgeWebChromeClient grants it at runtime); this plugin only answers the
 * gating question "does this device have a front camera?" — without opening the
 * camera or prompting — so the UI can hide the mirror button when there is none.
 *
 * hasFrontCamera() resolves { present: boolean }. Never rejects.
 */
@CapacitorPlugin(name = "PalMirror")
public class PalMirrorPlugin extends Plugin {

    @PluginMethod
    public void hasFrontCamera(PluginCall call) {
        boolean present = false;
        try {
            CameraManager manager =
                (CameraManager) getContext().getSystemService(Context.CAMERA_SERVICE);
            if (manager != null) {
                for (String id : manager.getCameraIdList()) {
                    Integer facing = manager.getCameraCharacteristics(id)
                        .get(CameraCharacteristics.LENS_FACING);
                    if (facing != null && facing == CameraCharacteristics.LENS_FACING_FRONT) {
                        present = true;
                        break;
                    }
                }
            }
        } catch (Exception e) {
            present = false;   // camera service unavailable → treat as no front cam
        }
        JSObject ret = new JSObject();
        ret.put("present", present);
        call.resolve(ret);
    }
}
